File: kb_languages/python/knowledge_base/kb_python/postgres/data_structures/kb_rpc_client.py
import time
import uuid
import json
from datetime import datetime, timezone
import psycopg2
from psycopg2.extras import execute_values
from psycopg2 import OperationalError
from psycopg2 import errors
from psycopg2 import sql
import psycopg2.extras
psycopg2.extras.register_uuid()
from psycopg2 import errors
class KB_RPC_Client:
    """
    A class to handle the RPC client for the knowledge base.
    """
    def __init__(self, kb_search,database):
        self.kb_search = kb_search
        self.conn = self.kb_search.conn
        self.cursor = self.kb_search.cursor 
        self.base_table = f"{database}_rpc_client"
        
    def find_rpc_client_id(self,kb=None,node_name=None, properties=None, node_path=None):
        """
        Find the node id for a given node name, properties, node path, and data.
        """
      
        result = self.find_node_ids(kb,node_name, properties, node_path)
        if len(result) == 0:
            raise ValueError(f"No node found matching path parameters: {node_name}, {properties}, {node_path}")
        if len(result) > 1:
            raise ValueError(f"Multiple nodes found matching path parameters: {node_name}, {properties}, {node_path}")
        return result
    
    def find_rpc_client_ids(self, kb=None, node_name=None, properties=None, node_path=None):
        """
        Find the node id for a given node name, properties, node path.
        """
       
        self.kb_search.clear_filters()
        self.kb_search.search_label("KB_RPC_CLIENT_FIELD")
        if kb is not None:
            self.kb_search.search_kb(kb)
        if node_name is not None:
            self.kb_search.search_name(node_name)
        if properties is not None:
            for key in properties:
                self.kb_search.search_property_value(key, properties[key])
        if node_path is not None:
            self.kb_search.search_path(node_path)
        node_ids = self.kb_search.execute_query()
        
        if node_ids is None:
            raise ValueError(f"No node found matching path parameters: {node_name}, {properties}, {node_path}")
        if len(node_ids) == 0:
            raise ValueError(f"No node found matching path parameters: {node_name}, {properties}, {node_path}")
        return node_ids
    
    def find_rpc_client_keys(self, key_data):
        """
        Extract key values from key_data.
        """
        return_values = []
        for key in key_data:
            return_values.append(key['path'])
        return return_values    
    
    
            
    def find_free_slots(self, client_path):
        """
        Find the number of free slots (records with is_new_result=FALSE) for a given client_path.
        This is a point-in-time snapshot and may change immediately after reading.
        
        Args:
            client_path (str): LTree compatible path for client
            
        Returns:
            int: Number of free slots available at the time of query
            
        Raises:
            Exception: If no records exist for the specified client_path
        """
        try:
            with self.conn.cursor() as cursor:
                # Use READ COMMITTED isolation level for consistent snapshot
                cursor.execute("SET TRANSACTION ISOLATION LEVEL READ COMMITTED")
                
                query = sql.SQL("""
                    SELECT 
                        COUNT(*) as total_records,
                        COUNT(*) FILTER (WHERE is_new_result = FALSE) as free_slots
                    FROM {table} 
                    WHERE client_path = %s
                """).format(table=sql.Identifier(self.base_table))
                
                cursor.execute(query, (client_path,))
                result = cursor.fetchone()
                
                total_records = result[0]
                free_slots = result[1]
                
                if total_records == 0:
                    raise Exception(f"No records found for client_path: {client_path}")
                
                return free_slots
                
        except psycopg2.Error as e:
            raise Exception(f"Database error when finding free slots: {str(e)}")
                  

    def find_queued_slots(self, client_path):
        """
        Find the number of queued slots (records with is_new_result=TRUE) for a given client_path.
        This is a point-in-time snapshot for monitoring purposes.
        
        Args:
            client_path (str): LTree compatible path for client
            
        Returns:
            int: Number of queued slots available for the client_path
            
        Raises:
            Exception: If no records exist for the specified client_path
        """
        try:
            with self.conn.cursor() as cursor:
                # Single query to get both total and queued slots
                query = sql.SQL("""
                    SELECT 
                        COUNT(*) as total_records,
                        COUNT(*) FILTER (WHERE is_new_result = TRUE) as queued_slots
                    FROM {table} 
                    WHERE client_path = %s
                """).format(table=sql.Identifier(self.base_table))
                
                cursor.execute(query, (client_path,))
                result = cursor.fetchone()
                
                total_records = result[0]
                queued_slots = result[1]
                
                if total_records == 0:
                    raise Exception(f"No records found for client_path: {client_path}")
                
                return queued_slots
                
        except psycopg2.Error as e:
            raise Exception(f"Database error when finding queued slots: {str(e)}")
        
            
    def peak_and_claim_reply_data(self,
                                client_path: str,
                                max_retries: int = 3,
                                retry_delay: float = 1.0
                                ) -> tuple[int, dict]:
        """
        Atomically fetch and mark the next available reply as processed.

        Args:
            client_path (str): ltree path of the client
            max_retries (int): number of retry attempts if row is locked
            retry_delay (float): delay in seconds between retries

        Returns:
            tuple[int, dict]: (record ID, remaining data as dict)

        Raises:
            Exception: If no new result is found for client_path
            RuntimeError: If lock could not be acquired within retries
        """
        attempt = 0
        table_ident = sql.Identifier(*self.base_table.split('.'))

        while attempt < max_retries:
            try:
                with self.conn.cursor() as cur:
                    update_query = sql.SQL("""
                        UPDATE {table}
                        SET is_new_result = FALSE
                        WHERE id = (
                            SELECT id
                            FROM {table}
                            WHERE client_path = %s
                            AND is_new_result = TRUE
                            ORDER BY response_timestamp ASC
                            FOR UPDATE SKIP LOCKED
                            LIMIT 1
                        )
                        RETURNING *
                    """).format(table=table_ident)

                    cur.execute(update_query, (client_path,))
                    row = cur.fetchone()

                    if row:
                        col_names = [desc.name for desc in cur.description]
                        data = dict(zip(col_names, row))
                        
                        self.conn.commit()
                        return  data

                    # Check if any matching unclaimed rows exist
                    check_query = sql.SQL("""
                        SELECT EXISTS (
                            SELECT 1 FROM {table}
                            WHERE client_path = %s AND is_new_result = TRUE
                        )
                    """).format(table=table_ident)

                    cur.execute(check_query, (client_path,))
                    exists = cur.fetchone()[0]

                    if not exists:
                        return None

                    self.conn.rollback()
                    attempt += 1
                    time.sleep(retry_delay)

            except errors.LockNotAvailable:
                self.conn.rollback()
                attempt += 1
                time.sleep(retry_delay)

        raise RuntimeError(f"Could not lock a new-reply row after {max_retries} attempts")
        
        
 

    def clear_reply_queue(self,
                        client_path: str,
                        max_retries: int = 3,
                        retry_delay: float = 1.0
                        ) -> int:
        """
        Clear the reply queue by resetting records matching the specified client_path.

        For each matching record:
        - Sets a unique UUID for request_id
        - Sets server_path equal to client_path
        - Resets response_payload to empty JSON object
        - Updates response_timestamp to current UTC time
        - Sets is_new_result to FALSE

        Includes record locking with retries to handle concurrent access.

        Args:
            client_path (ltree value): The client path to match for clearing records
            max_retries (int): Maximum number of retries for acquiring the lock
            retry_delay (float): Delay in seconds between retry attempts

        Returns:
            int: Number of records updated
        """
        attempt = 0
        while attempt < max_retries:
            try:
                with self.conn.cursor() as cur:
                    cur.execute("BEGIN;")

                    # Dynamically construct the SELECT ... FOR UPDATE statement
                    select_query = sql.SQL("""
                        SELECT id
                        FROM {table}
                        WHERE client_path = %s
                        FOR UPDATE NOWAIT
                    """).format(table=sql.Identifier(*self.base_table.split('.')))

                    cur.execute(select_query, (client_path,))
                    rows = cur.fetchall()

                    if not rows:
                        cur.execute("COMMIT;")
                        return 0

                    updated = 0
                    for (row_id,) in rows:
                        new_uuid = str(uuid.uuid4())

                        # Dynamically construct the UPDATE statement
                        update_query = sql.SQL("""
                            UPDATE {table}
                            SET
                                request_id         = %s,
                                server_path        = %s,
                                response_payload   = %s,
                                response_timestamp = NOW(),
                                is_new_result      = FALSE
                            WHERE id = %s
                        """).format(table=sql.Identifier(*self.base_table.split('.')))

                        cur.execute(update_query, (
                            new_uuid,
                            client_path,
                            json.dumps({}),
                            row_id
                        ))
                        updated += cur.rowcount

                    cur.execute("COMMIT;")
                    return updated

            except errors.LockNotAvailable:
                self.conn.rollback()
                attempt += 1
                time.sleep(retry_delay)

        raise RuntimeError(f"Could not acquire lock after {max_retries} retries")

    def push_and_claim_reply_data(self, client_path, request_uuid, server_path,
                                rpc_action, transaction_tag, reply_data,
                                max_retries=3, retry_delay=1):
        """
        Atomically claim and update the earliest matching record with is_new_result=FALSE.
        
        Args:
            client_path (str): LTree client path
            request_uuid (str): Request UUID
            server_path (str): LTree server path
            rpc_action (str): RPC action name
            transaction_tag (str): Transaction tag
            reply_data (dict): Reply data (to be stored as JSON)
            max_retries (int): Max retries on lock conflict
            retry_delay (float): Delay between retries
        
        Raises:
            Exception: On failure after all retries or if no matching record found.
        """
        attempt = 0
        last_error = None
        table_ident = sql.Identifier(*self.base_table.split('.'))

        while attempt <= max_retries:
            try:
                with self.conn.cursor() as cur:
                    # Begin transaction
                    cur.execute("BEGIN;")

                    # Combine SELECT and UPDATE atomically with locking
                    query = sql.SQL("""
                        WITH candidate AS (
                            SELECT id
                            FROM {table}
                            WHERE client_path = %s
                            AND is_new_result = FALSE
                            ORDER BY response_timestamp ASC
                            FOR UPDATE SKIP LOCKED
                            LIMIT 1
                        )
                        UPDATE {table}
                        SET request_id        = %s,
                            server_path       = %s,
                            rpc_action        = %s,
                            transaction_tag   = %s,
                            response_payload  = %s,
                            is_new_result     = TRUE,
                            response_timestamp = CURRENT_TIMESTAMP
                        FROM candidate
                        WHERE {table}.id = candidate.id
                        RETURNING {table}.id
                    """).format(table=table_ident)

                    cur.execute(query, (
                        client_path,
                        request_uuid,
                        server_path,
                        rpc_action,
                        transaction_tag,
                        json.dumps(reply_data)
                    ))

                    result = cur.fetchone()
                    if not result:
                        self.conn.rollback()
                        raise Exception("No available record with is_new_result=FALSE found")

                    self.conn.commit()
                    return  # success

            except (OperationalError, DatabaseError) as e:
                self.conn.rollback()
                last_error = e
                attempt += 1
                if attempt > max_retries:
                    raise Exception(f"Failed after {max_retries} retries: {str(last_error)}")
                time.sleep(retry_delay)

            except Exception as e:
                self.conn.rollback()
                raise e

        raise Exception(f"Failed after {max_retries} retries: {str(last_error)}")

    def list_waiting_jobs(self, client_path=None):
        """
        List all rows where is_new_result is TRUE, optionally filtered by client_path.

        Args:
            client_path (str, optional): If provided, filter results to this client_path

        Returns:
            list: A list of dictionaries, each containing the data for one waiting job

        Raises:
            Exception: If a database error occurs
        """
        try:
            with self.conn.cursor() as cursor:
                table_ident = sql.Identifier(*self.base_table.split('.'))
                
                if client_path is None:
                    query = sql.SQL("""
                        SELECT id, request_id, client_path, server_path, 
                            response_payload, response_timestamp, is_new_result
                        FROM {table}
                        WHERE is_new_result = TRUE
                        ORDER BY response_timestamp ASC
                    """).format(table=table_ident)
                    params = ()
                else:
                    query = sql.SQL("""
                        SELECT id, request_id, client_path, server_path, 
                            response_payload, response_timestamp, is_new_result
                        FROM {table}
                        WHERE is_new_result = TRUE AND client_path = %s
                        ORDER BY response_timestamp ASC
                    """).format(table=table_ident)
                    params = (client_path,)

                cursor.execute(query, params)
                column_names = [desc[0] for desc in cursor.description]
                records = cursor.fetchall()

                result = []
                for record in records:
                    record_dict = dict(zip(column_names, record))

                    if record_dict.get('request_id') is not None:
                        record_dict['request_id'] = str(record_dict['request_id'])

                    if isinstance(record_dict.get('response_timestamp'), datetime):
                        record_dict['response_timestamp'] = record_dict['response_timestamp'].isoformat()

                    for path_key in ('client_path', 'server_path'):
                        if record_dict.get(path_key) is not None:
                            record_dict[path_key] = str(record_dict[path_key])

                    result.append(record_dict)

                return result

        except psycopg2.Error as e:
            raise Exception(f"Database error when listing waiting jobs: {str(e)}")

File: python_programs_and_containers/system_control/file_loader/test_files/kb_go/postgres/data_structures/data_structures_module/KBRPCClient.go
package data_structures_module


import (
	"database/sql"
	"encoding/json"
	"fmt"
	//"strings"
	"time"

	"github.com/google/uuid"
	//"github.com/lib/pq"
)

// KBRPCClient handles RPC client operations for the knowledge base
type KBRPCClient struct {
	KBSearch  *KBSearch
	conn      *sql.DB
	BaseTable string
}

// ReplyData represents a reply data record
type ReplyData struct {
	ID                int                    `json:"id"`
	RequestID         string                 `json:"request_id"`
	ClientPath        string                 `json:"client_path"`
	ServerPath        string                 `json:"server_path"`
	RPCAction         string                 `json:"rpc_action,omitempty"`
	TransactionTag    string                 `json:"transaction_tag,omitempty"`
	ResponsePayload   map[string]interface{} `json:"response_payload"`
	ResponseTimestamp time.Time              `json:"response_timestamp"`
	IsNewResult       bool                   `json:"is_new_result"`
}

// SlotCounts represents free and queued slot counts
type SlotCounts struct {
	TotalRecords int `json:"total_records"`
	FreeSlots    int `json:"free_slots"`
	QueuedSlots  int `json:"queued_slots"`
}

// NewKBRPCClient creates a new KBRPCClient instance
func NewKBRPCClient(kbSearch *KBSearch, database string) *KBRPCClient {
	return &KBRPCClient{
		KBSearch:  kbSearch,
		conn:      kbSearch.conn,
		BaseTable: fmt.Sprintf("%s_rpc_client", database),
	}
}

// FindRPCClientID finds a single RPC client id for given parameters
func (client *KBRPCClient) FindRPCClientID(kb *string, nodeName *string, properties map[string]interface{}, nodePath *string) (map[string]interface{}, error) {
	results, err := client.FindRPCClientIDs(kb, nodeName, properties, nodePath)
	if err != nil {
		return nil, err
	}

	if len(results) == 0 {
		return nil, fmt.Errorf("no node found matching path parameters: %v, %v, %v", nodeName, properties, nodePath)
	}
	if len(results) > 1 {
		return nil, fmt.Errorf("multiple nodes found matching path parameters: %v, %v, %v", nodeName, properties, nodePath)
	}

	return results[0], nil
}

// FindRPCClientIDs finds all RPC client ids matching the given parameters
func (client *KBRPCClient) FindRPCClientIDs(kb *string, nodeName *string, properties map[string]interface{}, nodePath *string) ([]map[string]interface{}, error) {
	client.KBSearch.ClearFilters()
	client.KBSearch.SearchLabel("KB_RPC_CLIENT_FIELD")

	if kb != nil {
		client.KBSearch.SearchKB(*kb)
	}
	if nodeName != nil {
		client.KBSearch.SearchName(*nodeName)
	}
	if properties != nil {
		for key, value := range properties {
			client.KBSearch.SearchPropertyValue(key, value)
		}
	}
	if nodePath != nil {
		client.KBSearch.SearchPath(*nodePath)
	}

	nodeIDs, err := client.KBSearch.ExecuteQuery()
	if err != nil {
		return nil, err
	}

	if len(nodeIDs) == 0 {
		return nil, fmt.Errorf("no node found matching path parameters: %v, %v, %v", nodeName, properties, nodePath)
	}

	return nodeIDs, nil
}

// FindRPCClientKeys extracts path values from RPC client query results
func (client *KBRPCClient) FindRPCClientKeys(keyData []map[string]interface{}) []string {
	returnValues := []string{}
	for _, key := range keyData {
		if path, ok := key["path"].(string); ok {
			returnValues = append(returnValues, path)
		}
	}
	return returnValues
}

// FindFreeSlots finds the number of free slots for a given client path
func (client *KBRPCClient) FindFreeSlots(clientPath string) (int, error) {
	query := fmt.Sprintf(`
		SELECT 
			COUNT(*) as total_records,
			COUNT(*) FILTER (WHERE is_new_result = FALSE) as free_slots
		FROM %s 
		WHERE client_path = $1
	`, client.BaseTable)

	var totalRecords, freeSlots int
	err := client.conn.QueryRow(query, clientPath).Scan(&totalRecords, &freeSlots)
	if err != nil {
		return 0, fmt.Errorf("database error when finding free slots: %v", err)
	}

	if totalRecords == 0 {
		return 0, fmt.Errorf("no records found for client_path: %s", clientPath)
	}

	return freeSlots, nil
}

// FindQueuedSlots finds the number of queued slots for a given client path
func (client *KBRPCClient) FindQueuedSlots(clientPath string) (int, error) {
	query := fmt.Sprintf(`
		SELECT 
			COUNT(*) as total_records,
			COUNT(*) FILTER (WHERE is_new_result = TRUE) as queued_slots
		FROM %s 
		WHERE client_path = $1
	`, client.BaseTable)

	var totalRecords, queuedSlots int
	err := client.conn.QueryRow(query, clientPath).Scan(&totalRecords, &queuedSlots)
	if err != nil {
		return 0, fmt.Errorf("database error when finding queued slots: %v", err)
	}

	if totalRecords == 0 {
		return 0, fmt.Errorf("no records found for client_path: %s", clientPath)
	}

	return queuedSlots, nil
}

// PeakAndClaimReplyData atomically fetches and marks the next available reply as processed
func (client *KBRPCClient) PeakAndClaimReplyData(clientPath string, maxRetries int, retryDelay time.Duration) (*ReplyData, error) {
	if maxRetries <= 0 {
		maxRetries = 3
	}
	if retryDelay <= 0 {
		retryDelay = time.Second
	}

	attempt := 0
	for attempt < maxRetries {
		tx, err := client.conn.Begin()
		if err != nil {
			return nil, fmt.Errorf("failed to begin transaction: %v", err)
		}

		updateQuery := fmt.Sprintf(`
			UPDATE %s
			SET is_new_result = FALSE
			WHERE id = (
				SELECT id
				FROM %s
				WHERE client_path = $1
				AND is_new_result = TRUE
				ORDER BY response_timestamp ASC
				FOR UPDATE SKIP LOCKED
				LIMIT 1
			)
			RETURNING *
		`, client.BaseTable, client.BaseTable)

		rows, err := tx.Query(updateQuery, clientPath)
		if err != nil {
			tx.Rollback()
			if isLockError(err) && attempt < maxRetries-1 {
				attempt++
				time.Sleep(retryDelay)
				continue
			}
			return nil, err
		}
		defer rows.Close()

		if !rows.Next() {
			// Check if any matching unclaimed rows exist
			checkQuery := fmt.Sprintf(`
				SELECT EXISTS (
					SELECT 1 FROM %s
					WHERE client_path = $1 AND is_new_result = TRUE
				)
			`, client.BaseTable)

			var exists bool
			err = tx.QueryRow(checkQuery, clientPath).Scan(&exists)
			if err != nil {
				tx.Rollback()
				return nil, err
			}

			tx.Rollback()
			if !exists {
				return nil, nil
			}

			attempt++
			time.Sleep(retryDelay)
			continue
		}

		// Scan the row
		columns, err := rows.Columns()
		if err != nil {
			tx.Rollback()
			return nil, err
		}

		values := make([]interface{}, len(columns))
		valuePointers := make([]interface{}, len(columns))
		for i := range columns {
			valuePointers[i] = &values[i]
		}

		if err := rows.Scan(valuePointers...); err != nil {
			tx.Rollback()
			return nil, err
		}

		// Build result map
		result := make(map[string]interface{})
		for i, col := range columns {
			result[col] = values[i]
		}

		if err := tx.Commit(); err != nil {
			return nil, err
		}

		return mapToReplyData(result), nil
	}

	return nil, fmt.Errorf("could not lock a new-reply row after %d attempts", maxRetries)
}

// ClearReplyQueue clears the reply queue by resetting records matching the specified client path
func (client *KBRPCClient) ClearReplyQueue(clientPath string, maxRetries int, retryDelay time.Duration) (int, error) {
	if maxRetries <= 0 {
		maxRetries = 3
	}
	if retryDelay <= 0 {
		retryDelay = time.Second
	}

	attempt := 0
	for attempt < maxRetries {
		tx, err := client.conn.Begin()
		if err != nil {
			return 0, fmt.Errorf("failed to begin transaction: %v", err)
		}

		// Select and lock records
		selectQuery := fmt.Sprintf(`
			SELECT id
			FROM %s
			WHERE client_path = $1
			FOR UPDATE NOWAIT
		`, client.BaseTable)

		rows, err := tx.Query(selectQuery, clientPath)
		if err != nil {
			tx.Rollback()
			if isLockError(err) && attempt < maxRetries-1 {
				attempt++
				time.Sleep(retryDelay)
				continue
			}
			return 0, err
		}

		var ids []int
		for rows.Next() {
			var id int
			if err := rows.Scan(&id); err != nil {
				rows.Close()
				tx.Rollback()
				return 0, err
			}
			ids = append(ids, id)
		}
		rows.Close()

		if len(ids) == 0 {
			tx.Commit()
			return 0, nil
		}

		// Update each record
		updated := 0
		updateQuery := fmt.Sprintf(`
			UPDATE %s
			SET
				request_id         = $1,
				server_path        = $2,
				response_payload   = $3,
				response_timestamp = NOW(),
				is_new_result      = FALSE
			WHERE id = $4
		`, client.BaseTable)

		for _, id := range ids {
			newUUID := uuid.New().String()
			emptyJSON, _ := json.Marshal(map[string]interface{}{})
			
			result, err := tx.Exec(updateQuery, newUUID, clientPath, string(emptyJSON), id)
			if err != nil {
				tx.Rollback()
				return 0, err
			}
			
			rowsAffected, _ := result.RowsAffected()
			updated += int(rowsAffected)
		}

		if err := tx.Commit(); err != nil {
			return 0, err
		}

		return updated, nil
	}

	return 0, fmt.Errorf("could not acquire lock after %d retries", maxRetries)
}

// PushAndClaimReplyData atomically claims and updates the earliest matching record
func (client *KBRPCClient) PushAndClaimReplyData(clientPath, requestUUID, serverPath, rpcAction, 
	transactionTag string, replyData map[string]interface{}, maxRetries int, retryDelay time.Duration) error {
	
	if maxRetries <= 0 {
		maxRetries = 3
	}
	if retryDelay <= 0 {
		retryDelay = time.Second
	}

	replyJSON, err := json.Marshal(replyData)
	if err != nil {
		return fmt.Errorf("failed to marshal reply data: %v", err)
	}

	var lastError error
	for attempt := 0; attempt <= maxRetries; attempt++ {
		tx, err := client.conn.Begin()
		if err != nil {
			lastError = err
			continue
		}

		query := fmt.Sprintf(`
			WITH candidate AS (
				SELECT id
				FROM %s
				WHERE client_path = $1
				AND is_new_result = FALSE
				ORDER BY response_timestamp ASC
				FOR UPDATE SKIP LOCKED
				LIMIT 1
			)
			UPDATE %s
			SET request_id        = $2,
				server_path       = $3,
				rpc_action        = $4,
				transaction_tag   = $5,
				response_payload  = $6,
				is_new_result     = TRUE,
				response_timestamp = CURRENT_TIMESTAMP
			FROM candidate
			WHERE %s.id = candidate.id
			RETURNING %s.id
		`, client.BaseTable, client.BaseTable, client.BaseTable, client.BaseTable)

		var id int
		err = tx.QueryRow(query, clientPath, requestUUID, serverPath, rpcAction, 
			transactionTag, string(replyJSON)).Scan(&id)
		
		if err != nil {
			tx.Rollback()
			if err == sql.ErrNoRows {
				lastError = fmt.Errorf("no available record with is_new_result=FALSE found")
			} else {
				lastError = err
			}
			
			if attempt < maxRetries {
				time.Sleep(retryDelay)
				continue
			}
			break
		}

		if err := tx.Commit(); err != nil {
			lastError = err
			if attempt < maxRetries {
				time.Sleep(retryDelay)
				continue
			}
			break
		}

		return nil
	}

	return fmt.Errorf("failed after %d retries: %v", maxRetries, lastError)
}

// ListWaitingJobs lists all rows where is_new_result is TRUE
func (client *KBRPCClient) ListWaitingJobs(clientPath *string) ([]ReplyData, error) {
	var query string
	var args []interface{}

	if clientPath == nil {
		query = fmt.Sprintf(`
			SELECT id, request_id, client_path, server_path, rpc_action, transaction_tag,
				response_payload, response_timestamp, is_new_result
			FROM %s
			WHERE is_new_result = TRUE
			ORDER BY response_timestamp ASC
		`, client.BaseTable)
	} else {
		query = fmt.Sprintf(`
			SELECT id, request_id, client_path, server_path, rpc_action, transaction_tag,
				response_payload, response_timestamp, is_new_result
			FROM %s
			WHERE is_new_result = TRUE AND client_path = $1
			ORDER BY response_timestamp ASC
		`, client.BaseTable)
		args = append(args, *clientPath)
	}

	rows, err := client.conn.Query(query, args...)
	if err != nil {
		return nil, fmt.Errorf("database error when listing waiting jobs: %v", err)
	}
	defer rows.Close()

	var results []ReplyData
	for rows.Next() {
		var rd ReplyData
		var requestID sql.NullString
		var rpcAction sql.NullString
		var transactionTag sql.NullString
		var payloadStr string

		err := rows.Scan(&rd.ID, &requestID, &rd.ClientPath, &rd.ServerPath, 
			&rpcAction, &transactionTag, &payloadStr, &rd.ResponseTimestamp, &rd.IsNewResult)
		if err != nil {
			return nil, err
		}

		if requestID.Valid {
			rd.RequestID = requestID.String
		}
		if rpcAction.Valid {
			rd.RPCAction = rpcAction.String
		}
		if transactionTag.Valid {
			rd.TransactionTag = transactionTag.String
		}

		// Parse JSON payload
		if err := json.Unmarshal([]byte(payloadStr), &rd.ResponsePayload); err != nil {
			// If JSON parsing fails, store as raw string
			rd.ResponsePayload = map[string]interface{}{"raw": payloadStr}
		}

		results = append(results, rd)
	}

	if err := rows.Err(); err != nil {
		return nil, err
	}

	return results, nil
}

// Helper functions

// mapToReplyData converts a map to ReplyData struct
func mapToReplyData(m map[string]interface{}) *ReplyData {
	rd := &ReplyData{}

	if id, ok := m["id"].(int64); ok {
		rd.ID = int(id)
	}
	if requestID, ok := m["request_id"].(string); ok {
		rd.RequestID = requestID
	}
	if clientPath, ok := m["client_path"].(string); ok {
		rd.ClientPath = clientPath
	}
	if serverPath, ok := m["server_path"].(string); ok {
		rd.ServerPath = serverPath
	}
	if rpcAction, ok := m["rpc_action"].(string); ok {
		rd.RPCAction = rpcAction
	}
	if transactionTag, ok := m["transaction_tag"].(string); ok {
		rd.TransactionTag = transactionTag
	}
	if timestamp, ok := m["response_timestamp"].(time.Time); ok {
		rd.ResponseTimestamp = timestamp
	}
	if isNew, ok := m["is_new_result"].(bool); ok {
		rd.IsNewResult = isNew
	}

	// Handle response payload
	if payloadStr, ok := m["response_payload"].(string); ok {
		var payload map[string]interface{}
		if err := json.Unmarshal([]byte(payloadStr), &payload); err == nil {
			rd.ResponsePayload = payload
		}
	}

	return rd
}

/*
// isLockError checks if an error is a lock-related error
func isLockError(err error) bool {
	if err == nil {
		return false
	}
	pqErr, ok := err.(*pq.Error)
	if !ok {
		return false
	}
	// 55P03 = lock_not_available
	return pqErr.Code == "55P03"
}
*/

